File: android_flutter_claude/transcripto/packages/data_crypto/lib/src/ciphers/vigenere_cipher.dart
import 'dart:convert';
import 'package:core_common/core_common.dart';
import 'package:domain/domain.dart';

/// Vigenère cipher implementation
class VigenereCipher {
  const VigenereCipher();
  
  /// Encrypts text using Vigenère cipher
  String encrypt(String input, CipherParams params) {
    final key = _getEffectiveKey(params);
    return _processText(input, key, true);
  }
  
  /// Decrypts text using Vigenère cipher
  String decrypt(String input, CipherParams params) {
    final key = _getEffectiveKey(params);
    return _processText(input, key, false);
  }
  
  String _getEffectiveKey(CipherParams params) {
    if (params.key == null || params.key!.trim().isEmpty) {
      throw InvalidKeyException('Vigenère cipher requires a key');
    }
    
    String effectiveKey = params.key!.trim();
    
    // Add salt to key if enabled
    if (params.useSalt && params.salt != null) {
      final saltString = utf8.decode(params.salt!);
      effectiveKey = effectiveKey + saltString;
    }
    
    // Normalize and filter only letters
    final normalizedKey = Utils.normalizeString(effectiveKey);
    final lettersOnly = normalizedKey
        .split('')
        .where((char) => Utils.isLetter(char))
        .join()
        .toUpperCase();
    
    if (lettersOnly.isEmpty) {
      throw InvalidKeyException('Key must contain at least one letter');
    }
    
    return lettersOnly;
  }
  
  String _processText(String input, String key, bool encrypt) {
    final buffer = StringBuffer();
    int keyIndex = 0;
    
    for (int i = 0; i < input.length; i++) {
      final char = input[i];
      
      if (Utils.isLetter(char)) {
        final isUppercase = Utils.isUppercase(char);
        final charCode = char.toUpperCase().codeUnitAt(0);
        final keyChar = key[keyIndex % key.length];
        final keyCode = keyChar.codeUnitAt(0);
        
        final shift = keyCode - 65; // A=0, B=1, etc.
        int resultCode;
        
        if (encrypt) {
          resultCode = ((charCode - 65 + shift) % 26) + 65;
        } else {
          resultCode = ((charCode - 65 - shift + 26) % 26) + 65;
        }
        
        final resultChar = String.fromCharCode(resultCode);
        buffer.write(isUppercase ? resultChar : resultChar.toLowerCase());
        keyIndex++;
      } else {
        buffer.write(char); // Keep non-alphabetic characters unchanged
      }
    }
    
    return buffer.toString();
  }
}

File: andorid_deepseek/data/crypto/src/main/java/com/example/transcripto/data/crypto/impl/VigenereCipher.kt
package com.example.transcripto.data.crypto.impl

import com.example.transcripto.domain.crypto.CipherMethod
import com.example.transcripto.domain.crypto.CipherMethod.CipherParams
import javax.inject.Inject

class VigenereCipher @Inject constructor() : CipherMethod.Vigenere() {
    override fun encrypt(input: String, params: CipherParams): String {
        require(params.key.isNotEmpty()) { "Key is required for Vigenère cipher" }
        val effectiveKey = normalizeKey(params.key + params.salt)
        return input.mapIndexed { i, char ->
            when {
                char in 'A'..'Z' -> {
                    val shift = effectiveKey[i % effectiveKey.length] - 'A'
                    'A' + (char - 'A' + shift).mod(26)
                }
                char in 'a'..'z' -> {
                    val shift = effectiveKey[i % effectiveKey.length] - 'a'
                    'a' + (char - 'a' + shift).mod(26)
                }
                else -> char
            }
        }.joinToString("")
    }

    override fun decrypt(input: String, params: CipherParams): String {
        require(params.key.isNotEmpty()) { "Key is required for Vigenère cipher" }
        val effectiveKey = normalizeKey(params.key + params.salt)
        return input.mapIndexed { i, char ->
            when {
                char in 'A'..'Z' -> {
                    val shift = effectiveKey[i % effectiveKey.length] - 'A'
                    'A' + (char - 'A' - shift).mod(26)
                }
                char in 'a'..'z' -> {
                    val shift = effectiveKey[i % effectiveKey.length] - 'a'
                    'a' + (char - 'a' - shift).mod(26)
                }
                else -> char
            }
        }.joinToString("")
    }

    private fun normalizeKey(key: String): String {
        return key.filter { it.isLetter() }.takeIf { it.isNotEmpty() } 
            ?: throw IllegalArgumentException("Key must contain at least one letter")
    }
}